src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -40,32 +42,52 @@ internal static class DocumentBasedFixAllProviderHelpers
         var workItemCount = fixAllKind == FixAllKind.CodeFix ? 3 : 2;
         progressTracker.AddItems(fixAllContexts.Length * workItemCount);
 
-        // Process each context one at a time, allowing us to dump any information we computed for each once done with it.
-        var currentSolution = solution;
-        foreach (var fixAllContext in fixAllContexts)
+        using var _1 = PooledDictionary<DocumentId, (SyntaxNode? node, SourceText? text)>.GetInstance(out var allContextsDocIdToNewRootOrText);
         {
-            Contract.ThrowIfFalse(fixAllContext.Scope is FixAllScope.Document or FixAllScope.Project
-                or FixAllScope.ContainingMember or FixAllScope.ContainingType);
-            currentSolution = await FixSingleContextAsync(currentSolution, fixAllContext, progressTracker, getFixedDocumentsAsync).ConfigureAwait(false);
-        }
+            // First, iterate over all contexts, and collect all the changes for each of them.  We'll be making a lot of
+            // calls to the remote server to compute diagnostics and changes.  So keep a single connection alive to it
+            // so we never resync or recompute anything.
+            using var _2 = await RemoteKeepAliveSession.CreateAsync(solution, originalFixAllContext.CancellationToken).ConfigureAwait(false);
```

Comment:
this is the first large scale operation we do.  where we process the entire initial solution to determine the first pass of documents changed.   for this, we want things pinned so nothing drops between local and oop.